nvidia_gpu_stats/src/main.rs

```diff
@@ -0,0 +1,356 @@
+
+            if gpu_in_use {
+                metrics.insert(
+                    format!("gpu.process.{}.enforcedPowerLimitWatts", di),
+                    json!(power_limit),
+                );
+                metrics.insert(
+                    format!("gpu.process.{}.powerPercent", di),
+                    json!(power_percent),
+                );
+            }
+        }
+
+        // Additional metrics. These may not be available on all devices.
+        // Not reported to the backend, but could be useful for debugging
+        // and may be added in the future.
+        let graphics_clock = device.clock_info(Clock::Graphics)?;
+        metrics.insert(format!("_gpu.{}.graphicsClock", di), json!(graphics_clock));
+
+        let mem_clock = device.clock_info(Clock::Memory)?;
+        metrics.insert(format!("_gpu.{}.memoryClock", di), json!(mem_clock));
+
+        let link_gen = device.current_pcie_link_gen()?;
+        metrics.insert(format!("_gpu.{}.pcieLinkGen", di), json!(link_gen));
+
+        if let Ok(link_speed) = device.pcie_link_speed().map(u64::from).map(|x| x * 1000000) {
+            metrics.insert(format!("_gpu.{}.pcieLinkSpeed", di), json!(link_speed));
+        }
+
+        let link_width = device.current_pcie_link_width()?;
+        metrics.insert(format!("_gpu.{}.pcieLinkWidth", di), json!(link_width));
+
+        let max_link_gen = device.max_pcie_link_gen()?;
+        metrics.insert(format!("_gpu.{}.maxPcieLinkGen", di), json!(max_link_gen));
+
+        let max_link_width = device.max_pcie_link_width()?;
+        metrics.insert(
+            format!("_gpu.{}.maxPcieLinkWidth", di),
+            json!(max_link_width),
+        );
+
+        let cuda_cores = device.num_cores()?;
+        metrics.insert(format!("_gpu.{}.cudaCores", di), json!(cuda_cores));
+
+        let architecture = device.architecture()?;
+        metrics.insert(
+            format!("_gpu.{}.architecture", di),
+            json!(format!("{:?}", architecture)),
+        );
+    }
+
+    Ok(GpuMetrics { metrics })
+}
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // Parse command-line arguments
+    let args = Args::parse();
+
+    // Initialize NVML
+    let nvml_result = nvml_wrapper::Nvml::init();
+
+    // Set up a flag to control the main sampling loop
+    let running = Arc::new(AtomicBool::new(true));
+    let r = running.clone();
+
+    // Set up signal handler for graceful shutdown
+    let mut signals = Signals::new(TERM_SIGNALS)?;
+    thread::spawn(move || {
+        for _sig in signals.forever() {
+            r.store(false, Ordering::Relaxed);
+            break;
+        }
+    });
+
+    // Main sampling loop. Will run until the parent process is no longer alive or a signal is received.
+    while running.load(Ordering::Relaxed) {
+        let sampling_start = Instant::now();
+        let timestamp = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
```

Comment:
this can also panic. It might be OK in this case; just add a (short) comment justifying it

Reply:
fixed with a default